# Problem 2: KL Divergence and Maximum Likelihood

## Part (a): Nonnegativity of KL Divergence

We need to prove that:
1. $D_{KL}(P||Q) \geq 0$ for all distributions $P$ and $Q$
2. $D_{KL}(P||Q) = 0$ if and only if $P = Q$

### Proof:

**Step 1: Applying Jensen's Inequality**

We start with the definition of KL divergence:
$$D_{KL}(P||Q) = \sum_{x \in X} P(x) \log \frac{P(x)}{Q(x)}$$

We can rewrite this as:
$$D_{KL}(P||Q) = \sum_{x \in X} P(x) \log \frac{P(x)}{Q(x)} = -\sum_{x \in X} P(x) \log \frac{Q(x)}{P(x)}$$

Let's define the function $f(t) = -\log(t)$ for $t > 0$.

**Key observation**: $f(t) = -\log(t)$ is strictly convex on $(0, \infty)$ because:
- $f'(t) = -\frac{1}{t}$
- $f''(t) = \frac{1}{t^2} > 0$ for all $t > 0$

Since $f''(t) > 0$, the function $f$ is strictly convex.

**Step 2: Applying Jensen's Inequality to KL Divergence**

Now we apply Jensen's inequality. Let $X$ be a random variable that takes values in $\mathcal{X}$ with probability distribution $P$. 

Consider the random variable $Y = \frac{Q(X)}{P(X)}$. Then:

$$E[Y] = E\left[\frac{Q(X)}{P(X)}\right] = \sum_{x \in \mathcal{X}} P(x) \cdot \frac{Q(x)}{P(x)} = \sum_{x \in \mathcal{X}} Q(x) = 1$$

Since $f(t) = -\log(t)$ is strictly convex, Jensen's inequality gives us:
$$E[f(Y)] \geq f(E[Y])$$

Substituting our expressions:
$$E\left[-\log\left(\frac{Q(X)}{P(X)}\right)\right] \geq -\log\left(E\left[\frac{Q(X)}{P(X)}\right]\right)$$

This becomes:
$$\sum_{x \in \mathcal{X}} P(x) \cdot \left(-\log\left(\frac{Q(x)}{P(x)}\right)\right) \geq -\log(1) = 0$$

Therefore:
$$D_{KL}(P||Q) = -\sum_{x \in \mathcal{X}} P(x) \log\left(\frac{Q(x)}{P(x)}\right) \geq 0$$

**Step 3: Equality Condition**

Now we need to prove that $D_{KL}(P||Q) = 0$ if and only if $P = Q$.

**Part 1: If $P = Q$, then $D_{KL}(P||Q) = 0$**

If $P(x) = Q(x)$ for all $x \in \mathcal{X}$, then:
$$D_{KL}(P||Q) = \sum_{x \in \mathcal{X}} P(x) \log \frac{P(x)}{Q(x)} = \sum_{x \in \mathcal{X}} P(x) \log \frac{P(x)}{P(x)} = \sum_{x \in \mathcal{X}} P(x) \log(1) = \sum_{x \in \mathcal{X}} P(x) \cdot 0 = 0$$

**Part 2: If $D_{KL}(P||Q) = 0$, then $P = Q$**

Since we used Jensen's inequality with the strictly convex function $f(t) = -\log(t)$, equality holds in Jensen's inequality if and only if the random variable $Y = \frac{Q(X)}{P(X)}$ is constant almost surely.

If $D_{KL}(P||Q) = 0$, then we have equality in Jensen's inequality:
$$E[f(Y)] = f(E[Y])$$

Since $f$ is strictly convex, this happens if and only if $Y = \frac{Q(X)}{P(X)}$ is constant for all $x$ where $P(x) > 0$.

Let this constant be $c$. Then $Q(x) = c \cdot P(x)$ for all $x$ where $P(x) > 0$.

Since both $P$ and $Q$ are probability distributions:
$$\sum_{x \in \mathcal{X}} Q(x) = \sum_{x \in \mathcal{X}} c \cdot P(x) = c \sum_{x \in \mathcal{X}} P(x) = c \cdot 1 = c$$

But we also know that $\sum_{x \in \mathcal{X}} Q(x) = 1$, so $c = 1$.

Therefore, $Q(x) = P(x)$ for all $x \in \mathcal{X}$, which means $P = Q$.

**Conclusion**

We have successfully proven both parts of the required statement:

1. **Nonnegativity**: $D_{KL}(P||Q) \geq 0$ for all probability distributions $P$ and $Q$
   - This follows directly from Jensen's inequality applied to the strictly convex function $f(t) = -\log(t)$

2. **Equality condition**: $D_{KL}(P||Q) = 0$ if and only if $P = Q$
   - Forward direction: If $P = Q$, then clearly $D_{KL}(P||Q) = 0$
   - Reverse direction: If $D_{KL}(P||Q) = 0$, then equality holds in Jensen's inequality, which (for strictly convex functions) occurs if and only if the random variable is constant, leading to $P = Q$

This completes the proof of the nonnegativity property of KL divergence. ∎

In [1]:
import numpy as np
import matplotlib.pyplot as plt

def kl_divergence(p, q):
    """Compute KL divergence D_KL(P||Q)"""
    # Avoid log(0) by adding small epsilon
    epsilon = 1e-10
    return np.sum(p * np.log((p + epsilon) / (q + epsilon)))

# Example 1: Two different distributions
p1 = np.array([0.5, 0.3, 0.2])
q1 = np.array([0.4, 0.4, 0.2])

kl1 = kl_divergence(p1, q1)
print(f"Example 1: P = {p1}, Q = {q1}")
print(f"D_KL(P||Q) = {kl1:.6f}")
print(f"Is D_KL(P||Q) >= 0? {kl1 >= 0}")
print()

# Example 2: Identical distributions
p2 = np.array([0.5, 0.3, 0.2])
q2 = np.array([0.5, 0.3, 0.2])

kl2 = kl_divergence(p2, q2)
print(f"Example 2: P = {p2}, Q = {q2}")
print(f"D_KL(P||Q) = {kl2:.10f}")
print(f"Is D_KL(P||Q) = 0? {abs(kl2) < 1e-10}")
print()

# Example 3: Very different distributions
p3 = np.array([0.9, 0.05, 0.05])
q3 = np.array([0.1, 0.45, 0.45])

kl3 = kl_divergence(p3, q3)
print(f"Example 3: P = {p3}, Q = {q3}")
print(f"D_KL(P||Q) = {kl3:.6f}")
print(f"Is D_KL(P||Q) >= 0? {kl3 >= 0}")

Example 1: P = [0.5 0.3 0.2], Q = [0.4 0.4 0.2]
D_KL(P||Q) = 0.025267
Is D_KL(P||Q) >= 0? True

Example 2: P = [0.5 0.3 0.2], Q = [0.5 0.3 0.2]
D_KL(P||Q) = 0.0000000000
Is D_KL(P||Q) = 0? True

Example 3: P = [0.9  0.05 0.05], Q = [0.1  0.45 0.45]
D_KL(P||Q) = 1.757780
Is D_KL(P||Q) >= 0? True


## Part (b): Chain Rule for KL Divergence

We need to prove that:
$$D_{KL}(P(X,Y)||Q(X,Y)) = D_{KL}(P(X)||Q(X)) + D_{KL}(P(Y|X)||Q(Y|X))$$

Given the definition:
$$D_{KL}(P(X|Y)||Q(X|Y)) = \sum_y P(y) \sum_x P(x|y) \log \frac{P(x|y)}{Q(x|y)}$$

This is the expected KL divergence between conditional distributions on $x$, where the expectation is over $y$.

### Proof:

**Step 1: Start with the left-hand side (LHS)**

We begin with the KL divergence between the joint distributions $P(X,Y)$ and $Q(X,Y)$:

$$D_{KL}(P(X,Y)||Q(X,Y)) = \sum_{x,y} P(x,y) \log \frac{P(x,y)}{Q(x,y)}$$

Using the chain rule of probability, we can factorize the joint distributions:
- $P(x,y) = P(x) \cdot P(y|x)$
- $Q(x,y) = Q(x) \cdot Q(y|x)$

Substituting these into the KL divergence:

$$D_{KL}(P(X,Y)||Q(X,Y)) = \sum_{x,y} P(x,y) \log \frac{P(x) \cdot P(y|x)}{Q(x) \cdot Q(y|x)}$$

**Step 2: Apply logarithm properties**

Using the property $\log \frac{ab}{cd} = \log \frac{a}{c} + \log \frac{b}{d}$, we can split the logarithm:

$$D_{KL}(P(X,Y)||Q(X,Y)) = \sum_{x,y} P(x,y) \left[\log \frac{P(x)}{Q(x)} + \log \frac{P(y|x)}{Q(y|x)}\right]$$

Distributing the sum:

$$D_{KL}(P(X,Y)||Q(X,Y)) = \sum_{x,y} P(x,y) \log \frac{P(x)}{Q(x)} + \sum_{x,y} P(x,y) \log \frac{P(y|x)}{Q(y|x)}$$

We now have two terms to analyze separately.

**Step 3: Simplify the first term**

Let's focus on the first term:

$$\text{Term 1} = \sum_{x,y} P(x,y) \log \frac{P(x)}{Q(x)}$$

Notice that $\log \frac{P(x)}{Q(x)}$ does not depend on $y$. We can factor it out of the inner sum:

$$\text{Term 1} = \sum_x \log \frac{P(x)}{Q(x)} \sum_y P(x,y)$$

The inner sum $\sum_y P(x,y) = P(x)$ by marginalization. Therefore:

$$\text{Term 1} = \sum_x P(x) \log \frac{P(x)}{Q(x)} = D_{KL}(P(X)||Q(X))$$

This is exactly the first term on the right-hand side (RHS) of the chain rule!

**Step 4: Simplify the second term**

Now let's focus on the second term:

$$\text{Term 2} = \sum_{x,y} P(x,y) \log \frac{P(y|x)}{Q(y|x)}$$

Using the chain rule of probability, we can write $P(x,y) = P(x) \cdot P(y|x)$:

$$\text{Term 2} = \sum_{x,y} P(x) \cdot P(y|x) \log \frac{P(y|x)}{Q(y|x)}$$

Rearranging the double sum:

$$\text{Term 2} = \sum_x P(x) \sum_y P(y|x) \log \frac{P(y|x)}{Q(y|x)}$$

Notice that the inner sum is exactly the KL divergence $D_{KL}(P(Y|X=x)||Q(Y|X=x))$ for a fixed value of $x$.

**Step 5: Recognize the conditional KL divergence**

The second term can be written as:

$$\text{Term 2} = \sum_x P(x) \cdot D_{KL}(P(Y|X=x)||Q(Y|X=x))$$

This is the **expected KL divergence** between the conditional distributions $P(Y|X)$ and $Q(Y|X)$, where the expectation is taken with respect to $P(X)$. 

By definition, this is exactly:

$$\text{Term 2} = D_{KL}(P(Y|X)||Q(Y|X))$$

This matches the second term on the RHS of the chain rule!

**Conclusion**

Combining both terms from Step 3 and Step 5:

$$D_{KL}(P(X,Y)||Q(X,Y)) = D_{KL}(P(X)||Q(X)) + D_{KL}(P(Y|X)||Q(Y|X))$$

This completes the proof of the chain rule for KL divergence. 

**Key insights:**
1. The chain rule for KL divergence mirrors the chain rule for probability (joint = marginal × conditional)
2. The first term captures the divergence between the marginal distributions of $X$
3. The second term captures the expected divergence between the conditional distributions of $Y$ given $X$
4. This decomposition shows that the total divergence can be split into a "marginal component" and a "conditional component"

This result is particularly useful in variational inference and information theory, where we often need to analyze how distributions differ both marginally and conditionally. ∎

In [1]:
import numpy as np

def kl_divergence_discrete(p, q):
    """Compute KL divergence D_KL(P||Q) for discrete distributions"""
    epsilon = 1e-10
    return np.sum(p * np.log((p + epsilon) / (q + epsilon)))

def kl_divergence_joint(p_xy, q_xy):
    """Compute KL divergence for joint distributions"""
    return kl_divergence_discrete(p_xy.flatten(), q_xy.flatten())

def kl_divergence_marginal(p_xy, q_xy):
    """Compute KL divergence for marginal distributions P(X) and Q(X)"""
    p_x = np.sum(p_xy, axis=1)  # Marginalize over Y
    q_x = np.sum(q_xy, axis=1)
    return kl_divergence_discrete(p_x, q_x)

def kl_divergence_conditional(p_xy, q_xy):
    """
    Compute expected KL divergence D_KL(P(Y|X)||Q(Y|X))
    This is the expectation over P(X) of D_KL(P(Y|X=x)||Q(Y|X=x))
    """
    p_x = np.sum(p_xy, axis=1, keepdims=True)  # P(X)
    q_x = np.sum(q_xy, axis=1, keepdims=True)  # Q(X)
    
    # Compute conditional distributions
    epsilon = 1e-10
    p_y_given_x = p_xy / (p_x + epsilon)  # P(Y|X)
    q_y_given_x = q_xy / (q_x + epsilon)  # Q(Y|X)
    
    # Expected KL divergence
    expected_kl = 0.0
    for i in range(p_xy.shape[0]):
        if p_x[i, 0] > epsilon:
            kl_conditional = kl_divergence_discrete(p_y_given_x[i, :], q_y_given_x[i, :])
            expected_kl += p_x[i, 0] * kl_conditional
    
    return expected_kl

# Example: Verify the chain rule
print("=" * 70)
print("Verifying Chain Rule for KL Divergence")
print("=" * 70)

# Define joint distribution P(X,Y) - 3x3 matrix
# Rows represent X values, columns represent Y values
P_XY = np.array([
    [0.15, 0.10, 0.05],  # P(X=0, Y=0), P(X=0, Y=1), P(X=0, Y=2)
    [0.20, 0.15, 0.10],  # P(X=1, Y=0), P(X=1, Y=1), P(X=1, Y=2)
    [0.10, 0.10, 0.05]   # P(X=2, Y=0), P(X=2, Y=1), P(X=2, Y=2)
])

# Define joint distribution Q(X,Y)
Q_XY = np.array([
    [0.12, 0.12, 0.08],
    [0.18, 0.18, 0.12],
    [0.08, 0.08, 0.04]
])

# Verify they are valid probability distributions
print(f"\nSum of P(X,Y): {np.sum(P_XY):.6f} (should be 1.0)")
print(f"Sum of Q(X,Y): {np.sum(Q_XY):.6f} (should be 1.0)")

# Compute LHS: D_KL(P(X,Y)||Q(X,Y))
lhs = kl_divergence_joint(P_XY, Q_XY)
print(f"\n{'Left-Hand Side:':<40}")
print(f"{'D_KL(P(X,Y)||Q(X,Y))':<40} = {lhs:.10f}")

# Compute RHS terms
term1 = kl_divergence_marginal(P_XY, Q_XY)
term2 = kl_divergence_conditional(P_XY, Q_XY)
rhs = term1 + term2

print(f"\n{'Right-Hand Side:':<40}")
print(f"{'D_KL(P(X)||Q(X))':<40} = {term1:.10f}")
print(f"{'D_KL(P(Y|X)||Q(Y|X))':<40} = {term2:.10f}")
print(f"{'Sum':<40} = {rhs:.10f}")

# Verify the chain rule
print(f"\n{'Verification:':<40}")
print(f"{'Difference (LHS - RHS)':<40} = {abs(lhs - rhs):.2e}")
print(f"{'Chain rule verified?':<40} {abs(lhs - rhs) < 1e-9}")

# Display marginal distributions
p_x_marginal = np.sum(P_XY, axis=1)
q_x_marginal = np.sum(Q_XY, axis=1)

print(f"\n{'Marginal Distributions:':<40}")
print(f"P(X) = {p_x_marginal}")
print(f"Q(X) = {q_x_marginal}")

print("\n" + "=" * 70)

Verifying Chain Rule for KL Divergence

Sum of P(X,Y): 1.000000 (should be 1.0)
Sum of Q(X,Y): 1.000000 (should be 1.0)

Left-Hand Side:                         
D_KL(P(X,Y)||Q(X,Y))                     = 0.0230167973

Right-Hand Side:                        
D_KL(P(X)||Q(X))                         = 0.0073819970
D_KL(P(Y|X)||Q(Y|X))                     = 0.0156348004
Sum                                      = 0.0230167973

Verification:                           
Difference (LHS - RHS)                   = 2.39e-11
Chain rule verified?                     True

Marginal Distributions:                 
P(X) = [0.3  0.45 0.25]
Q(X) = [0.32 0.48 0.2 ]



### Interpretation of the Numerical Example

The numerical verification above confirms the chain rule for KL divergence with high precision. Let's interpret the results:

**The Chain Rule Decomposition:**
- **Total divergence** $D_{KL}(P(X,Y)||Q(X,Y)) = 0.0230$ nats
  - This represents the total information loss or discrepancy when using $Q(X,Y)$ to approximate $P(X,Y)$

**Component Analysis:**
1. **Marginal component** $D_{KL}(P(X)||Q(X)) = 0.0074$ nats (32% of total)
   - This captures how much the marginal distributions of $X$ differ
   - From the example: $P(X) = [0.3, 0.45, 0.25]$ vs $Q(X) = [0.32, 0.48, 0.20]$
   
2. **Conditional component** $D_{KL}(P(Y|X)||Q(Y|X)) = 0.0156$ nats (68% of total)
   - This captures the expected divergence in how $Y$ depends on $X$ under the two distributions
   - This is the larger contributor in this example, indicating that the main difference between $P$ and $Q$ lies in their conditional structure rather than their marginals

**Key Takeaway:**
The chain rule allows us to decompose the divergence between joint distributions into interpretable components: one measuring differences in marginal behavior, and another measuring differences in conditional dependencies. This is particularly useful in:
- **Model comparison**: Understanding whether models differ more in their marginal predictions or conditional relationships
- **Variational inference**: Optimizing factorized approximations to complex distributions
- **Causal analysis**: Separating distributional shift in variables from changes in their dependencies

### **2(c) [5 points] KL and maximum likelihood**

Consider a density estimation problem, and suppose we are given a training set $\{x^{(i)}; i = 1, \dots, m\}$. Let the empirical distribution be $\hat{P}(x) = \frac{1}{m} \sum_{i=1}^m 1\{x^{(i)} = x\}$. ($\hat{P}$ is just the uniform distribution over the training set; i.e., sampling from the empirical distribution is the same as picking a random example from the training set).

Suppose we have some family of distributions $P_\theta$ parameterized by $\theta$. (If you like, think of $P_\theta(x)$ as an alternative notation for $P(x; \theta)$.) Prove that finding the maximum likelihood estimate for the parameter $\theta$ is equivalent to finding $P_\theta$ with minimal KL divergence from $\hat{P}$. I.e. prove:

$$ \text{arg min}_\theta D_{KL}(\hat{P} \| P_\theta) = \text{arg max}_\theta \sum_{i=1}^m \log P_\theta(x^{(i)}) $$

**Remark.** Consider the relationship between parts (b-c) and multi-variate Bernoulli Naive Bayes parameter estimation. In the Naive Bayes model we assumed $P_\theta$ is of the following form: $P_\theta(x, y) = p(y) \prod_{i=1}^n p(x_i|y)$. By the chain rule for KL divergence, we therefore have:

$$ D_{KL}(\hat{P} \| P_\theta) = D_{KL}(\hat{P}(y) \| p(y)) + \sum_{i=1}^n D_{KL}(\hat{P}(x_i|y) \| p(x_i|y)). $$

This shows that finding the maximum likelihood/minimum KL-divergence estimate of the parameters decomposes into $2n + 1$ independent optimization problems: One for the class priors $p(y)$, and one for each of the conditional distributions $p(x_i|y)$ for each feature $x_i$ given each of the two possible labels for $y$. Specifically, finding the maximum likelihood estimates for each of these problems individually results in also maximizing the likelihood of the joint distribution. (If you know what Bayesian networks are, a similar remark applies to parameter estimation for them).

### Proof:

We start with the definition of the KL divergence between the empirical distribution $\hat{P}$ and the model distribution $P_\theta$:

$$ D_{KL}(\hat{P} \| P_\theta) = \sum_{x} \hat{P}(x) \log \frac{\hat{P}(x)}{P_\theta(x)} $$

Using the property of logarithms $\log(a/b) = \log a - \log b$, we can expand this as:

$$ D_{KL}(\hat{P} \| P_\theta) = \sum_{x} \hat{P}(x) \log \hat{P}(x) - \sum_{x} \hat{P}(x) \log P_\theta(x) $$

Notice that the first term, $\sum_{x} \hat{P}(x) \log \hat{P}(x)$, depends only on the empirical distribution $\hat{P}$ (the data) and is independent of the parameter $\theta$. Let's denote this constant term as $H(\hat{P})$ (which is actually the negative entropy of $\hat{P}$).

$$ D_{KL}(\hat{P} \| P_\theta) = \text{constant} - \sum_{x} \hat{P}(x) \log P_\theta(x) $$

Now, let's substitute the definition of the empirical distribution $\hat{P}(x) = \frac{1}{m} \sum_{i=1}^m 1\{x^{(i)} = x\}$ into the second term:

$$ \sum_{x} \hat{P}(x) \log P_\theta(x) = \sum_{x} \left( \frac{1}{m} \sum_{i=1}^m 1\{x^{(i)} = x\} \right) \log P_\theta(x) $$

We can swap the order of summation:

$$ = \frac{1}{m} \sum_{i=1}^m \sum_{x} 1\{x^{(i)} = x\} \log P_\theta(x) $$

The inner sum over $x$ has only one non-zero term, which occurs when $x = x^{(i)}$. Therefore:

$$ \sum_{x} 1\{x^{(i)} = x\} \log P_\theta(x) = \log P_\theta(x^{(i)}) $$

Substituting this back:

$$ \sum_{x} \hat{P}(x) \log P_\theta(x) = \frac{1}{m} \sum_{i=1}^m \log P_\theta(x^{(i)}) $$

So the KL divergence expression becomes:

$$ D_{KL}(\hat{P} \| P_\theta) = \text{constant} - \frac{1}{m} \sum_{i=1}^m \log P_\theta(x^{(i)}) $$

Now, we want to find the $\theta$ that minimizes this KL divergence:

$$ \text{arg min}_\theta D_{KL}(\hat{P} \| P_\theta) = \text{arg min}_\theta \left( \text{constant} - \frac{1}{m} \sum_{i=1}^m \log P_\theta(x^{(i)}) \right) $$

Since the constant term does not affect the minimization, and minimizing a negative function is equivalent to maximizing the positive function (and the factor $1/m$ is positive and constant):

$$ \text{arg min}_\theta D_{KL}(\hat{P} \| P_\theta) = \text{arg max}_\theta \sum_{i=1}^m \log P_\theta(x^{(i)}) $$

The term on the right is exactly the log-likelihood of the data under the model $P_\theta$.

**Conclusion:**
Minimizing the KL divergence between the empirical distribution and the model distribution is mathematically equivalent to maximizing the likelihood (or log-likelihood) of the data. This provides a fundamental information-theoretic justification for the Maximum Likelihood Estimation (MLE) principle.

In [ ]:
import numpy as np
import matplotlib.pyplot as plt

def verify_mle_kl_equivalence():
    print("=" * 70)
    print("Numerical Verification: MLE vs Minimum KL Divergence")
    print("=" * 70)

    # 1. Generate synthetic data
    # True distribution: Bernoulli(p=0.7)
    np.random.seed(42)
    true_p = 0.7
    m = 1000  # Number of samples
    data = np.random.binomial(1, true_p, m)
    
    # 2. Compute Empirical Distribution P_hat
    # For binary data, P_hat is just defined by the fraction of 1s
    p_hat_1 = np.mean(data)
    p_hat_0 = 1 - p_hat_1
    P_hat = np.array([p_hat_0, p_hat_1])
    
    print(f"True parameter p: {true_p}")
    print(f"Empirical p_hat:  {p_hat_1:.4f}")
    
    # 3. Define parameter grid for theta (probability of 1)
    thetas = np.linspace(0.01, 0.99, 100)
    
    log_likelihoods = []
    kl_divergences = []
    
    for theta in thetas:
        # Model distribution P_theta
        P_theta = np.array([1 - theta, theta])
        
        # Calculate Log-Likelihood: sum(log P_theta(x_i))
        # We can use the counts directly:
        # LL = count(0) * log(1-theta) + count(1) * log(theta)
        ll = np.sum(data == 0) * np.log(1 - theta) + np.sum(data == 1) * np.log(theta)
        log_likelihoods.append(ll)
        
        # Calculate KL Divergence: D_KL(P_hat || P_theta)
        # D_KL = sum P_hat(x) * log(P_hat(x) / P_theta(x))
        kl = np.sum(P_hat * np.log(P_hat / P_theta))
        kl_divergences.append(kl)
        
    # Find optima
    max_ll_idx = np.argmax(log_likelihoods)
    min_kl_idx = np.argmin(kl_divergences)
    
    best_theta_ll = thetas[max_ll_idx]
    best_theta_kl = thetas[min_kl_idx]
    
    print(f"\nTheta maximizing Log-Likelihood: {best_theta_ll:.4f}")
    print(f"Theta minimizing KL Divergence:    {best_theta_kl:.4f}")
    print(f"Difference: {abs(best_theta_ll - best_theta_kl):.10f}")
    
    # Plotting
    fig, ax1 = plt.subplots(figsize=(10, 6))
    
    color = 'tab:blue'
    ax1.set_xlabel('Theta (Parameter)')
    ax1.set_ylabel('Log-Likelihood', color=color)
    ax1.plot(thetas, log_likelihoods, color=color, linewidth=2, label='Log-Likelihood')
    ax1.tick_params(axis='y', labelcolor=color)
    ax1.axvline(best_theta_ll, color=color, linestyle='--', alpha=0.5)
    
    ax2 = ax1.twinx()  # instantiate a second axes that shares the same x-axis
    
    color = 'tab:red'
    ax2.set_ylabel('KL Divergence', color=color)  # we already handled the x-label with ax1
    ax2.plot(thetas, kl_divergences, color=color, linewidth=2, label='KL Divergence')
    ax2.tick_params(axis='y', labelcolor=color)
    ax2.axvline(best_theta_kl, color=color, linestyle='--', alpha=0.5)
    
    plt.title('Equivalence of Max Likelihood and Min KL Divergence')
    fig.tight_layout()  # otherwise the right y-label is slightly clipped
    plt.grid(True, alpha=0.3)
    plt.show()

verify_mle_kl_equivalence()